sageresearch-app-sdk/src/main/java/org/sagebionetworks/research/sageresearch/viewmodel/ScheduleRepository.kt

```diff
@@ -176,20 +226,206 @@ open class ScheduleRepository(context: Context) {
+        // Or the user is offline and did a task without having synced the study's schedules.
+        // Which, that is okay, the s3 upload will still take place, but we won't be able to update the schedule on bridge.
+        // Side effects for this will be, no study reports or finishedOn status for the task will show in the UI.
+        // TODO: mdephillips 9/14/2018 message the user there will be no history?
+        val guid = scheduleTaskRunUuidMap[taskRunUuid] ?: return
+
+        // If we previously registered a guid with a taskRunUuid, we should be able to find the schedule in the
+        // the database at this point based on the schedule guid
+        findSchedule(guid, Action1 { schedule ->
+            // TODO: mdephillips 9/14/2018 in past apps, here we would set client data from TaskResult
+            // TODO: mdephillips 9/14/2018 instead use TaskResult to generate study report for schedule
+            schedule.startedOn = Instant.ofEpochMilli(taskResult.startDate?.time ?: Date().time)
+            schedule.finishedOn = Instant.ofEpochMilli(taskResult.endDate?.time ?: Date().time)
+            cacheSchedule(schedule)
+            updateScheduleToBridge(schedule)
+        }, Action1 {
+            // TODO: mdephillips 9/14/2018 message the user there will be no history?
+        })
+    }
+
+    /**
+     * Private function should only be accessed through this class as a re-usable way to update
+     * a schedule on bridge.
+     */
+    private fun updateScheduleToBridge(schedule: ScheduledActivityEntity) {
+        val bridgeSchedule = schedule.clientWritableCopy()
+        updateActivityOnBridge(bridgeSchedule, Action1 {
+            schedule.needsSyncedToBridge = false
+            cacheSchedule(schedule)
+        }, Action1 {
+            scheduleUpdateFailed(schedule, it)
+        })
+    }
+
+    /**
+     * Encapsulate db call functionality into it own function for providing custom mock behavior in tests,
+     * and also to remove redundant threading code throughout the class.
+     */
+    @VisibleForTesting
+    protected open fun findSchedule(
+            scheduleGuid: String,
+            onNext: Action1<ScheduledActivityEntity>,
+            onError: Action1<Throwable>) {
+
+        scheduleRepoSubscriptions.add(Observable.just(scheduleDao)
+                .subscribeOn(Schedulers.io())
+                .subscribe({ dao ->
+                    dao.activity(scheduleGuid).firstOrNull()?.let {
+                        scheduleOnMain(it, onNext, onError)
+                    } ?: run {
+                        val error = Throwable("No schedule found in DB with guid $scheduleGuid")
+                        scheduleOnMain(error, onError)
+                    }
+                }, {
+                    scheduleOnMain(it, onError)
+                }))
+    }
+
+    /**
+     * Helper function to schedule calls on the main thread
+     */
+    @VisibleForTesting
+    protected open fun scheduleOnMain(
```

Comment:
Is there a better way to do this?

Reply:
Is this what you're thinking of?

```observable
.operatorX // called on IO thread
.subscribeOn(Schedulers.io()) // subscribe thread actually can only be set once
.operatorY // called on IO thread
.observeOn(AndroidSchedulers.mainThread()) //switches  subsequent operators to main thread
.operatorZ // called on main thread

Reply:
Hmmm it's not letting me call observeOn after a subscribeOn

Reply:
checked out this branch (on your updated PR) taking a look